### PR TITLE
Type-level literals and Nat-dependent vector/matrix

### LINKED ISSUE #178: Are nat-dependent types possible?
I am writing a matrix type and a matrix multiplication function. What I would like is to have types `Matrix<n,m>` indexed by positive integers `n`, `m` so that the multiplication has signature 

`Matrix<n,m> -> Matrix<m,p> -> Matrix<n,p>` 

(so trying to multiply matrices of incompatible dimensions is a type error).

I have seen an implementation of this in F# (https://www.youtube.com/watch?v=3zdlQ_HjKl4 ) but the approach is quite hacky... Is it possible to do this neatly with F#+?

### LINKED ISSUE #201: What nat-dependent Vector & Matrix operations should be implemented?
Related: #178, #192 

I'm currently trying to add type-safe vector & matrix types to F#+ but I'm not sure what to implement.

Here are the ones I can think of:

* Basic collection operations (such as `map` and `filter`) in a type-safe manner (e.g. `item` indexed with type-level natural).
* Addition and subtraction
* Scalar multiplication
* dot product and 3D cross product for vectors
* matrix product for matrices
* constants such as zero, one, unit vectors, identity matrix

but I can imagine many other necessary operations are missing.

Can anyone provide me a list of other operations which can take advantage of type-level naturals and/or should be implemented?

### PR DESCRIPTION
**Pull request by cannorin**
Closes #178.
Closes #201.

This PR attempts to simulate nat-dependent vector and matrix types through type-level literals.

Things to be discussed:
- Operator naming (currently adding a suffix `^` to avoid problems with globally-defined operators such as `=` and `+`)
- Where to place these things (currently in a separated assembly for faster debugging)
- To be added?

TODO:

- [x] Implement type-level booleans and naturals
- [x] Add type provider for type-level naturals
- [x] Implement nat-dependent vector types
- [x] Implement nat-dependent matrix types as well
- [x] Add some "tests" (which is run by compiling it) if needed

### CODE REVIEW

**Comment by gusty** [comment]
Any clue of why is it failing the test only in appveyor?

**Comment by wallymathieu** [comment]
Could it be something like the build configuration of the solution? That the project does not get built?

**Comment by cannorin** [comment]
Can I turn this into Ready for Review (although actually not) to get a T-shirt on Hacktoberfest? :laughing:

**Comment by cannorin** [comment]
Vectors and matrices are now Applicative.

**Comment by wallymathieu** [comment]
I can do a small fix and push it to this branch

**Comment by cannorin** [comment]
@wallymathieu Nice! Can you send me a PR to that branch on my fork, or is it better to move the whole branch to this repo and reopen this PR?

**Comment by wallymathieu** [comment]
I can send you a PR

**Comment by wallymathieu** [comment]
Did you see the PR? It's essentially trivial changes in order to register project in solution

**Comment by cannorin** [comment]
sorry my internet connection is cursed these days and I can barely git push and see emails. I'll go to univ today and check the PR with a wifi.

**Comment by cannorin** [comment]
Merged :bowing_man:

**Comment by wallymathieu** [comment]
Looks like there is some need for travisconfiguration as well.

**Comment by wallymathieu** [comment]
Looks like paket restore is done after build why you have this PR https://github.com/cannorin/FSharpPlus/pull/2

**Comment by cannorin** [comment]
Merged! Thanks :smile:

**Comment by wallymathieu** [comment]
Looks like it's green as well! Awesome work @cannorin !

**Comment by cannorin** [comment]
Can we add a reference to MathNet.Numerics? Writing a SVD algorithm by ourselves seems not a good idea.

**Comment by gusty** [comment]
Hmmm, that's tricky.

Why do we need that algorithm?
To me that's an indication that we're going too far.

**Comment by wallymathieu** [comment]
Perhaps it should be a separate delivery?

**Comment by cannorin** [comment]
Fair point. For now I'll leave the matrix-related algorighms for future/separated work.

**Comment by cannorin** [comment]
Hmm paket restore on Travis stopped working?

**Comment by cannorin** [comment]
> Possible Performance degradation, V3 was not working: One or more errors occurred. (One or more errors occurred. (Could not load resources from 'https://api.nuget.org/v3/index.json'))

**Comment by cannorin** [comment]
btw we have several things to consider:

* Are we fine with that operator namings (such as `+^`, `*^`, etc...) ?
* Do we merge the FSharpPlus.TypeLits assembly into the FSharpPlus assembly?
* If not, how do we support `Zero` and `One` operation on `Matrix.zero` and `Matrix.identity`? (currently `LanguagePrimitives.GenericFoo` is used)

**Comment by cannorin** [comment]
Thanks @wallymathieu!

File: src/FSharpPlus.TypeLits/Matrix.fs [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+
+    static member inline CountTuple (_: Tuple<_>, _: CountTuple) = S Z
+    static member inline CountTuple ((_, _), _: CountTuple) = S Z |> S
+    static member inline CountTuple ((_, _, _), _: CountTuple) = S Z |> S |> S
+    static member inline CountTuple ((_, _, _, _), _: CountTuple) = S Z |> S |> S |> S
+    static member inline CountTuple ((_, _, _, _, _), _: CountTuple) = S Z |> S |> S |> S |> S
+    static member inline CountTuple ((_, _, _, _, _, _), _: CountTuple) = S Z |> S |> S |> S |> S |> S
+    static member inline CountTuple ((_, _, _, _, _, _, _), _: CountTuple) = S Z |> S |> S |> S |> S |> S |> S
+    
+  type TupleToList =
```
Comment (gusty) [inline src/FSharpPlus.TypeLits/Matrix.fs]: I think functions like this could be moved to the Tuple.fs and be reused from there.
Maybe they can even go in a separate PR.

Comment (cannorin) [reply src/FSharpPlus.TypeLits/Matrix.fs]: I'm going to merge FSharpPlus.TypeLits.fsproj into FSharpPlus.fsproj, then it would be good to move them to Tuple.fs while doing that

Comment (gusty) [reply src/FSharpPlus.TypeLits/Matrix.fs]: Nice, please go ahead.

**Comment by cannorin** [comment]
Merged the project. Also, `Vector.zero`, `Matrix.zero`, and `Matrix.identity` now uses `Zero.Invoke` and `One.Invoke`.

**Comment by cannorin** [comment]
oops, the merge just broke the fsproj terribly

**Comment by wallymathieu** [comment]
Aa, well, I can perhaps look into it during the weekend

**Comment by gusty** [comment]
It looks like it's because of the new structure.
Now most files are grouped in folder, and some of them are split in different files.

**Comment by wallymathieu** [comment]
It just needed a bit of reordering.

**Comment by cannorin** [comment]
Don't merge this yet, I have one more thing to try implementing

File: src/FSharpPlus/Data/Matrix.fs [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+
+// Items : 'Item[ 'Column, 'Row ]
+[<Struct; StructuredFormatDisplay("{Items}")>]
+type Matrix< 'Item, 'Row, 'Column > = private { Items: 'Item[,] } with
+  [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
+  member this.UnsafeGet (i, j) = this.Items.[i, j]
+  [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
+  static member UnsafeCreate (_row: 'm, _column: 'n, items: _[,]) : Matrix<_, 'm, 'n> =
+    { Items = items }
+  interface System.Collections.Generic.IReadOnlyCollection<'Item> with
```
Comment (gusty) [inline src/FSharpPlus/Data/Matrix.fs]: Can we comment out these interfaces implementation for now, until we merge #207 ?

Comment (cannorin) [reply src/FSharpPlus/Data/Matrix.fs]: OK!

**Comment by cannorin** [comment]
Ready to merge :tada:

File: src/FSharpPlus/Data/Matrix.fs (hunk 2 above)
Comment (gusty) [reply src/FSharpPlus/Data/Matrix.fs]: Actually we can leave the interface's members, there's no harm at all.
I propose just leave commented out the interfaces declaration, that will suffice for now.

File: src/FSharpPlus/Data/Matrix.fs [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+  static member inline ( @* ) (v1, v2) = Vector.directProduct v1 v2
+  /// cross product
+  static member inline ( %* ) (v1, v2) = Vector.vectorProduct3 v1 v2
+  /// hadamard (element-wise) product
+  static member inline ( * ) (v1: Vector<_, 'n>, v2: Vector<_, 'n>) = Vector.map2 (*) v1 v2
+  static member inline ( / ) (v1: Vector<_, 'n>, v2: Vector<_, 'n>) = Vector.map2 (/) v1 v2
+  static member inline ( * ) (v: Vector<'a, 'n>, s: 'a) = Vector.map (fun x -> x * s) v
+  static member inline ( * ) (s: 'a, v: Vector<'a, 'n>) = Vector.map (fun x -> x * s) v
+  static member inline ( / ) (v: Vector<'a, 'n>, s: 'a) = Vector.map (fun x -> x / s) v
+  static member inline ( / ) (s: 'a, v: Vector<'a, 'n>) = Vector.map (fun x -> s / x) v
```
Comment (gusty) [inline src/FSharpPlus/Data/Matrix.fs]: Here I see problem.
It's my understanding that we shouldn't add overloads for binary operations accepting a scalar and a vector, by doing so we're complicating type inference to a point where it's not decidable.
In order to multiply (or divide) by a scalar there are many other ways, the easiest one being to use generic literals (see the example I added in the docs).

Comment (cannorin) [reply src/FSharpPlus/Data/Matrix.fs]: OK, we can use `result` as in the doc anyway.

Comment (gusty) [reply src/FSharpPlus/Data/Matrix.fs]: Yes, `return`, `map` / `<!>`, applicative math operators or generic literals. But not automatically.
I think generic literals is the easiest, it requires only one additional keystroke.

File: src/FSharpPlus/Data/Matrix.fs (hunk 2 above)
Comment (gusty) [reply src/FSharpPlus/Data/Matrix.fs]: Now that I think, since F# has not implicit interfaces, it would create a breaking change, as those methods would become available without casting to the interface.

So better to leave the whole commented out as you already did.
